rollout/trafficrouting/nginx/nginx.go

```diff
@@ -0,0 +1,223 @@
+				hasStableServiceBackendRule = true
+				desiredCanaryIngress.Spec.Rules[ir].HTTP.Paths[ip].Backend.ServiceName = canaryServiceName
+			}
+		}
+	}
+	if !hasStableServiceBackendRule {
+		return nil, errors.New(fmt.Sprintf("ingress `%s` has no rules using service %s backend", stableIngressName, stableServiceName))
+	}
+
+	if desiredCanaryIngress.Annotations == nil {
+		desiredCanaryIngress.Annotations = map[string]string{}
+	}
+
+	// Process additional annotations, prepend annotationPrefix unless supplied. We are keeping all the annotations
+	// from the stableIngress since the controller automatically ignores most of them anyway:
+	// See: https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#canary
+	for k, v := range r.cfg.Rollout.Spec.Strategy.Canary.TrafficRouting.Nginx.AdditionalIngressAnnotations {
+		if !strings.HasPrefix(k, annotationPrefix) {
+			k = fmt.Sprintf("%s/%s", annotationPrefix, k)
+		}
+		desiredCanaryIngress.Annotations[k] = v
+	}
+	// Always set `canary` and `canary-weight` - `canary-by-header` and `canary-by-cookie`, if set,  will always take precedence
+	desiredCanaryIngress.Annotations[fmt.Sprintf("%s/canary", annotationPrefix)] = "true"
+	desiredCanaryIngress.Annotations[fmt.Sprintf("%s/canary-weight", annotationPrefix)] = fmt.Sprintf("%d", desiredWeight)
+
+	return desiredCanaryIngress, nil
+}
+
+// compareCanaryIngresses compares the current canaryIngress with the desired one and returns a patch
+func compareCanaryIngresses(current *extensionsv1beta1.Ingress, desired *extensionsv1beta1.Ingress) ([]byte, bool, error) {
+	// only compare Spec, Annotations, and Labels
+	return diff.CreateTwoWayMergePatch(
+		&extensionsv1beta1.Ingress{
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: current.Annotations,
+				Labels:      current.Labels,
+			},
+			Spec: current.Spec,
+		},
+		&extensionsv1beta1.Ingress{
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: desired.Annotations,
+				Labels:      desired.Labels,
+			},
+			Spec: desired.Spec,
+		}, extensionsv1beta1.Ingress{})
+}
+
+// Reconcile modifies Nginx Ingress resources to reach desired state
+func (r *Reconciler) Reconcile(desiredWeight int32) error {
+	stableIngressName := r.cfg.Rollout.Spec.Strategy.Canary.TrafficRouting.Nginx.StableIngress
+	canaryIngressName := fmt.Sprintf("%s-canary", stableIngressName)
+
+	// Check if stable ingress exists, error if it does not
+	stableIngress, err := r.cfg.Client.ExtensionsV1beta1().Ingresses(r.cfg.Rollout.Namespace).Get(stableIngressName, metav1.GetOptions{})
+	if err != nil {
+		r.log.WithField(logutil.IngressKey, stableIngressName).Errorf("Error retrieving ingress: %v", err)
+		return err
+	}
+
+	// Check if canary ingress exists, determines whether we later call Create() or Update()
+	canaryIngress, err := r.cfg.Client.ExtensionsV1beta1().Ingresses(r.cfg.Rollout.Namespace).Get(canaryIngressName, metav1.GetOptions{})
```

Comment:
I'd rather avoid making an API server call on each reconciliation if possible. I think we can achieve this with the following approach:

- Get Canary Ingress from IngressLister
-  If canary Ingress not found
   - Create Canary Ingress
   - If successful
     - return nil
   - If isAlreadyExists
     - Get Canary Ingress from API server
- Compare desired and current Canary Ingress and apply Patch

Reply:
While I wouldn't say it's a one to one comparison, we take a similar approach with ReplicaSets: https://github.com/argoproj/argo-rollouts/blob/master/rollout/sync.go#L56-L207

Reply:
I think one of the issues I'm hitting here is that the canaryIngress isn't actually reconciled in the Ingress controller, but instead in the Rollouts controller. How would you propose I proceed? I could move all the canary ingress creation and reconciliation out of the Rollouts controller and into the ingress controller (and have the rollouts controller basically work off the assumption that the canaryIngress exists, and only update the annotation).

Reply:
I'm still leaning towards having the Ingress creation in the Rollouts controller otherwise we will have to coordinate the Ingress and Rollout controllers. For example, if the rollout controller doesn't find the canary Ingress, did an error occur when the Ingress was created or does the rollout controller need to wait a little longer? That's hard for the Rollout controller to know without information passed to it or the controller stores information from the previous reconciliation. I'm worried that is going to increase the scope of this PR significantly.

Reply:
Why not require users to create a secondary Ingress on their own? Users are already required to create a secondary Service for their canary.

Reply:
This (reducing API calls) should be addressed in https://github.com/argoproj/argo-rollouts/pull/426/commits/86e1169913e54d714a1dd32a4866f18af3ebb237